test/micro/org/openjdk/bench/java/lang/foreign/CriticalCalls.java

```diff
@@ -0,0 +1,113 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.java.lang.foreign;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.foreign.SymbolLookup;
+
+import java.lang.invoke.MethodHandle;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.concurrent.TimeUnit;
+
+import static java.lang.foreign.ValueLayout.*;
+
+@BenchmarkMode(Mode.AverageTime)
+@Warmup(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@State(org.openjdk.jmh.annotations.Scope.Thread)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@Fork(value = 3, jvmArgsAppend = { "--enable-native-access=ALL-UNNAMED" })
+public class CriticalCalls {
+
+    static final MethodHandle PINNED;
+    static final MethodHandle NOT_PINNED;
+
+    static {
+        System.loadLibrary("CriticalCalls");
+        SymbolLookup lookup = SymbolLookup.loaderLookup();
+
+        MemorySegment sumIntsSym = lookup.find("sum_ints").get();
+        FunctionDescriptor sumIntsDesc = FunctionDescriptor.of(JAVA_INT, ADDRESS, JAVA_INT);
+
+        PINNED = Linker.nativeLinker().downcallHandle(
+            sumIntsSym,
+            sumIntsDesc,
+            Linker.Option.critical(true));
+        NOT_PINNED = Linker.nativeLinker().downcallHandle(
+            sumIntsSym,
+            sumIntsDesc,
+            Linker.Option.critical(false));
+    }
+
+    @Param({"100", "10000", "1000000"})
+    int size;
+
+    int[] arr;
+    SegmentAllocator recycler;
+
+    @Setup
+    public void setup() {
+        arr = new int[size];
+        for (int i = 0; i < size; i++) {
+            arr[i] = i;
+        }
+
+        recycler = SegmentAllocator.prefixAllocator(Arena.ofAuto().allocate(JAVA_INT, arr.length));
+    }
+
+    @Benchmark
+    public int callPinned() throws Throwable {
+        return (int) PINNED.invokeExact(MemorySegment.ofArray(arr), arr.length);
+    }
+
+    @Benchmark
+    public int callNotPinned() throws Throwable {
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment nativeArr = arena.allocate(JAVA_INT, arr.length);
```

Comment:
Good point. I made this benchmark a while ago, before we had the more optimized allocate variants

Reply:
This had a pretty big impact, actually. Especially on the larger sizes:

```
Benchmark                     (size)  Mode  Cnt       Score       Error  Units
CriticalCalls.callNotPinned      100  avgt   30      84.818 �     0.729  ns/op
CriticalCalls.callNotPinned    10000  avgt   30    2966.918 �    39.898  ns/op
CriticalCalls.callNotPinned  1000000  avgt   30  952864.052 � 34996.156  ns/op
CriticalCalls.callPinned         100  avgt   30      30.640 �     0.173  ns/op
CriticalCalls.callPinned       10000  avgt   30    2241.403 �    35.473  ns/op
CriticalCalls.callPinned     1000000  avgt   30  221152.247 �  1697.968  ns/op
CriticalCalls.callRecycled       100  avgt   30      40.205 �     0.458  ns/op
CriticalCalls.callRecycled     10000  avgt   30    2845.316 �    13.331  ns/op
CriticalCalls.callRecycled   1000000  avgt   30  287752.178 �  2322.382  ns/op
```

Reply:
I also notice that the non pinned variant of the `100` benchmark is slow compared to the others. This might be due to try with resources inhibiting scalarization. I suggest to call Arena::close explicitly in that benchmark and repeat the test.

Reply:
Not sure... The `callNotPinned` variant is meant as a typical use case where the native segment needs to be allocated. I think TWR belongs in that typical use-case.

This is really about measuring the difference between 2 idiomatic code patterns. If non-scalarization is part of one of those patterns, then I think that is something that should be included in the measurement.